pkg/releaselock/controller.go

```diff
@@ -0,0 +1,203 @@
+					klog.Errorf("Failed to sync lock info for configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+				}
+			}
+		}, defaultSyncPeriod)
+	}
+
+	rl, err := resourcelock.New(
+		resourcelock.LeasesResourceLock,
+		c.leaderElectionNamespace,
+		leaseName,
+		nil,
+		c.client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity: c.id,
+		})
+	if err != nil {
+		klog.Fatalf("Error creating resourcelock: %v", err)
+	}
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: c.leaseDuration,
+		RenewDeadline: c.renewDeadline,
+		RetryPeriod:   c.retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				klog.Fatalf("%s no longer the leader, staying inactive.", c.id)
+			},
+		},
+	})
+}
+
+func (c *LockReleaseController) syncLockInfo(ctx context.Context, cm *corev1.ConfigMap) error {
+	nodeName, err := util.GKENodeNameFromConfigMap(cm)
+	if err != nil {
+		klog.Errorf("Failed to get GKE node name from configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+		return err
+	}
+
+	klog.Infof("Getting GKE Node %s from API Server", nodeName)
+	node, err := c.client.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+	if err != nil && !apiError.IsNotFound(err) {
+		klog.Errorf("Failed to get node %s: %v", nodeName, err)
+		return err
+	}
+
+	data := cm.DeepCopy().Data
+	recentCM := cm.DeepCopy()
+	for key, filestoreIP := range data {
+		_, _, _, _, gkeNodeID, gkeNodeInternalIP, err := util.ParseConfigMapKey(key)
+		if err != nil {
+			klog.Errorf("Failed to parse configmap key %s: %v", key, err)
+			continue
+		}
+		klog.Infof("Verifying GKE node %s with nodeId %s nodeInternalIP %s exists or not", nodeName, gkeNodeID, gkeNodeInternalIP)
+		if c.verifyNodeExists(node, gkeNodeID, gkeNodeInternalIP) {
+			klog.Infof("GKE node %s with nodeId %s nodeInternalIP %s still exists in API server, skip lock info reconciliation", nodeName, gkeNodeID, gkeNodeInternalIP)
+			continue
+		}
+		klog.Infof("GKE node %s with nodeId %s nodeInternalIP %s no longer exists, releasing lock for GKE node IP %s Filestore IP %s", nodeName, gkeNodeID, gkeNodeInternalIP, gkeNodeInternalIP, filestoreIP)
+		if err := ReleaseLock(filestoreIP, gkeNodeInternalIP); err != nil {
+			klog.Errorf("Failed to release lock: %v", err)
+			continue
+		}
+		recentCM, err = util.RemoveKeyFromConfigMap(ctx, recentCM, key, c.client)
+		if err != nil {
+			klog.Errorf("Failed to remove key %s from configmap %s/%s: %v", key, cm.Namespace, cm.Name, err)
+		}
+	}
+	return nil
+}
+
+// verifyNodeExists validates if the given node object has the exact nodeID, and nodeInternalIP.
+func (c *LockReleaseController) verifyNodeExists(node *corev1.Node, expectedNodeID, expectedNodeInternalIP string) bool {
+	if node == nil {
+		return false
+	}
+	if node.Annotations == nil {
+		klog.Warningf("Node %s is unhealthy: node.annotations is nil", node.Name)
```

Comment:
have you tested this is how it shows up in a unhealthy node? same question for the missing gceInstanceIDKey?

Reply:
To be honest, I never see any cases so far when the node misses `gceInstanceIDKey or annotation` in common shutdown scenarios, the instance id always exists. The AI for me is to keep tracking the node unhealthy symptoms. 
But since it's an annotation stamping on a k8s object, I think these checks are still required. The [GKE monitor server](https://source.corp.google.com/piper///depot/google3/cloud/kubernetes/monitoring/node/health.go;l=213) also tracks this annotation to report gke node health.

Reply:
As discussed we do not want to wait for a node to report healthy again to trigger lock release. all we care about is the gce instance id to be removed

Reply:
done